wolfssl: add benchmark, enable CPU crypto instructions

cotequeiroz opened the pull request:
This enables AES & SHA CPU instructions for compatible armv8, and x86_64 architectures.

The package was marked non-shared, since the arm CPUs may or may not have crypto extensions enabled based on licensing, and there's no run-time detection, unlike x86_64.

Signed-off-by: Eneas U de Queiroz <cotequeiroz@gmail.com>

---

Discussion about this started in #9697.

I'm initially marking this as draft, as different targets need to be tested.  I've briefly searched the net for indications of crypto support (`aes pmull sha1 sha2` cpu flags).  Here are my findings (crossed out targets mean they do not support crypto asm, we should still confirm this info; checked boxes mean that the information has been confirmed):

- [ ] armvirt/64: should be OK

- [x] ~bcm27xx/bcm2710~: [no-aes: `fp asimd evtstrm crc32`](https://github.com/ThomasKaiser/sbc-bench/blob/5f295de52798366c5a1baa136c5403743f94fa76/sbc-bench.sh#L2307-L2309)

- [x] ~bcm27xx/bcm2711~: [no aes: `fp asimd evtstrm crc32`](https://github.com/ThomasKaiser/sbc-bench/blob/5f295de52798366c5a1baa136c5403743f94fa76/sbc-bench.sh#L2400-L2402)

- [ ] bcm4908: [`fp asimd evtstrm aes pmull sha1 sha2 crc32`](https://forum.openwrt.org/t/building-openwrt-for-64bit-socs/13331/12)
- [ ] layerscape/armv8_64b:
       	LS1012A: [datasheet mentions "ARMv8 cryptography extensions"](https://community.nxp.com/pwmxy87654/attachments/pwmxy87654/cw-qoriq/6139/1/LS1012A_Rev0.pdf)
       	LS1046A: [`fp asimd evtstrm aes pmull sha1 sha2 crc32 cpuid`](https://www.avnet.com/wps/wcm/connect/onesite/ca6887b0-90f0-4d67-b68d-0bd2a52950bf/Layerscape+FRWY-LS1046A+BSP+User+Guide.pdf?MOD=AJPERES&CVID=nCJ6uqe&CVID=nCJ6uqe&attachment=false&id=1622108828554)
       	LS1043A: [datasheet mentions "ARMv8 cryptography extensions"](https://www.mouser.com/datasheet/2/302/nxp_phgl-s-a0004270254-1-1750392.pdf)
       	LS1088A: [datasheet mentions "ARMv8 cryptography extensions"](https://community.nxp.com/pwmxy87654/attachments/pwmxy87654/imx-processors/161806/1/NXP%20LS1088A-1386069.pdf)

- [x] mediatek/mt7622: Tested locally; `fp asimd evtstrm aes pmull sha1 sha2 crc32 cpuid`

- [x] mvebu/cortexa53: [`fp asimd evtstrm aes pmull sha1 sha2 crc32`](https://github.com/ThomasKaiser/sbc-bench/blob/5f295de52798366c5a1baa136c5403743f94fa76/sbc-bench.sh#L2392-L2395).

- [x] mvebu/cortexa72:  confirmed by @robimarko.

- [x] octeontx: `CONFIG_CRYPTO_AES_ARM64_CE=y` in [target/linux/octeontx/config-5.10](https://github.com/openwrt/openwrt/blob/810b48e79389a54b9b332fa2834360602a9734f3/target/linux/octeontx/config-5.10#L98-L100)

- [x] rockchip/armv8:
     rk3328: cortex-a53: [`fp asimd evtstrm aes pmull sha1 sha2 crc32`](https://github.com/ThomasKaiser/sbc-bench/blob/5f295de52798366c5a1baa136c5403743f94fa76/sbc-bench.sh#L2329-L2331)
       	       	rk3399: cortex-a72: [`fp asimd evtstrm aes pmull sha1 sha2 crc32`](https://github.com/ThomasKaiser/sbc-bench/blob/5f295de52798366c5a1baa136c5403743f94fa76/sbc-bench.sh#L2425-L2428) 

- [ ] sunxi/cortexa53: [`fp asimd evtstrm aes pmull sha1 sha2 crc32`](https://github.com/ThomasKaiser/sbc-bench/blob/5f295de52798366c5a1baa136c5403743f94fa76/sbc-bench.sh#L2325-L2328)

cotequeiroz committed:
wolfssl: enable CPU crypto instructions

This enables AES & SHA CPU instructions for compatible armv8, and x86_64
architectures.  Add this to the hardware acceleration choice, since they
can't be enabled at the same time.

The package was marked non-shared, since the arm CPUs may or may not
have crypto extensions enabled based on licensing; bcm27xx does not
enable them.  There is no run-time detection of this for arm.

NOTE:
Should this be backported to a release branch, it must be done shortly
before a new minor release, because the change to nonshared will remove
libwolfssl from the shared packages, but the nonshared are only built in
a subsequent release!

Signed-off-by: Eneas U de Queiroz <cotequeiroz@gmail.com>

robimarko commented:
Thanks for working on this,
I can confirm that Mvebu A53 and A72 both support the crypto extensions according to the datasheet.
A72 also supports CRC extension but that is not relevant here.

cotequeiroz commented:
It appears right now, that we can reverse the selection logic: all subtargets support the crypto extensions, except bcm27xx.

robimarko commented:
I will check whether I have acess to Octeon TX datasheet as it being enabled in the kernel doesnt meant that the CPU supports it

cotequeiroz commented:
I know.  It would just be ironical if the only `aarch64` physical target that had the CE drivers selected two weeks ago, would side with bcm27xx in not supporting the extensions.

cotequeiroz commented:
Here is a `diff --side` of the results of the benchmark running on Linksys E8450 with default options:
```
-------------------------------------------------------------  	-------------------------------------------------------------
 wolfSSL version 5.2.0 	       	       	       	       	       	 wolfSSL version 5.2.0
-------------------------------------------------------------  	-------------------------------------------------------------
wolfCrypt Benchmark (block bytes 1048576, min 1.0 sec each)    	wolfCrypt Benchmark (block bytes 1048576, min 1.0 sec each)
RNG                 20 MB took 1.012 seconds,   19.756 MB/s   |	RNG                 85 MB took 1.013 seconds,   83.947 MB/s
AES-128-CBC-enc     30 MB took 1.023 seconds,   29.324 MB/s   |	AES-128-CBC-enc    680 MB took 1.001 seconds,  679.150 MB/s
AES-128-CBC-dec     30 MB took 1.107 seconds,   27.106 MB/s   |	AES-128-CBC-dec    645 MB took 1.006 seconds,  641.272 MB/s
AES-192-CBC-enc     30 MB took 1.168 seconds,   25.674 MB/s   |	AES-192-CBC-enc    595 MB took 1.000 seconds,  594.775 MB/s
AES-192-CBC-dec     25 MB took 1.070 seconds,   23.356 MB/s   |	AES-192-CBC-dec    590 MB took 1.001 seconds,  589.401 MB/s
AES-256-CBC-enc     25 MB took 1.099 seconds,   22.745 MB/s   |	AES-256-CBC-enc    550 MB took 1.002 seconds,  548.705 MB/s
AES-256-CBC-dec     25 MB took 1.209 seconds,   20.675 MB/s   |	AES-256-CBC-dec    520 MB took 1.005 seconds,  517.427 MB/s
AES-128-GCM-enc     20 MB took 1.134 seconds,   17.637 MB/s   |	AES-128-GCM-enc    435 MB took 1.007 seconds,  431.904 MB/s
AES-128-GCM-dec     20 MB took 1.136 seconds,   17.601 MB/s   |	AES-128-GCM-dec    200 MB took 1.003 seconds,  199.401 MB/s
AES-192-GCM-enc     20 MB took 1.237 seconds,   16.170 MB/s   |	AES-192-GCM-enc    395 MB took 1.003 seconds,  393.652 MB/s
AES-192-GCM-dec     20 MB took 1.235 seconds,   16.195 MB/s   |	AES-192-GCM-dec    195 MB took 1.020 seconds,  191.101 MB/s
AES-256-GCM-enc     20 MB took 1.331 seconds,   15.031 MB/s   |	AES-256-GCM-enc    380 MB took 1.005 seconds,  378.251 MB/s
AES-256-GCM-dec     15 MB took 1.002 seconds,   14.973 MB/s   |	AES-256-GCM-dec    190 MB took 1.019 seconds,  186.399 MB/s
GMAC Table 4-bit    45 MB took 1.006 seconds,   44.742 MB/s   |	GMAC Table 4-bit   322 MB took 1.001 seconds,  321.761 MB/s
AES-128-ECB-enc     28 MB took 1.000 seconds,   27.776 MB/s   |	AES-128-ECB-enc    202 MB took 1.000 seconds,  202.354 MB/s
AES-128-ECB-dec     27 MB took 1.000 seconds,   27.024 MB/s   |	AES-128-ECB-dec    206 MB took 1.000 seconds,  206.428 MB/s
AES-192-ECB-enc     25 MB took 1.000 seconds,   25.487 MB/s   |	AES-192-ECB-enc    184 MB took 1.000 seconds,  184.133 MB/s
AES-192-ECB-dec     24 MB took 1.000 seconds,   23.623 MB/s   |	AES-192-ECB-dec    188 MB took 1.000 seconds,  187.800 MB/s
AES-256-ECB-enc     23 MB took 1.000 seconds,   22.757 MB/s   |	AES-256-ECB-enc    170 MB took 1.000 seconds,  170.011 MB/s
AES-256-ECB-dec     21 MB took 1.000 seconds,   20.843 MB/s   |	AES-256-ECB-dec    173 MB took 1.000 seconds,  173.132 MB/s
AES-128-CTR         30 MB took 1.038 seconds,   28.897 MB/s   |	AES-128-CTR        885 MB took 1.002 seconds,  883.201 MB/s
AES-192-CTR         30 MB took 1.187 seconds,   25.265 MB/s   |	AES-192-CTR        815 MB took 1.002 seconds,  813.554 MB/s
AES-256-CTR         25 MB took 1.118 seconds,   22.366 MB/s   |	AES-256-CTR        760 MB took 1.004 seconds,  756.799 MB/s
AES-CCM-Enc         15 MB took 1.022 seconds,   14.681 MB/s   |	AES-CCM-Enc        100 MB took 1.001 seconds,   99.885 MB/s
AES-CCM-Dec         15 MB took 1.025 seconds,   14.633 MB/s   |	AES-CCM-Dec        105 MB took 1.041 seconds,  100.828 MB/s
ARC4                70 MB took 1.064 seconds,   65.789 MB/s   |	ARC4                70 MB took 1.056 seconds,   66.286 MB/s
CHACHA              80 MB took 1.059 seconds,   75.550 MB/s   |	CHACHA             300 MB took 1.006 seconds,  298.357 MB/s
CHA-POLY            60 MB took 1.077 seconds,   55.717 MB/s   |	CHA-POLY           205 MB took 1.016 seconds,  201.849 MB/s
3DES                10 MB took 1.871 seconds,    5.345 MB/s   |	3DES                10 MB took 1.940 seconds,    5.154 MB/s
MD5                135 MB took 1.018 seconds,  132.667 MB/s   |	MD5                135 MB took 1.015 seconds,  133.022 MB/s
POLY1305           240 MB took 1.018 seconds,  235.738 MB/s   |	POLY1305           610 MB took 1.001 seconds,  609.575 MB/s
SHA                110 MB took 1.005 seconds,  109.452 MB/s   |	SHA                115 MB took 1.026 seconds,  112.102 MB/s
SHA-224             50 MB took 1.043 seconds,   47.923 MB/s   |	SHA-224            605 MB took 1.001 seconds,  604.342 MB/s
SHA-256             50 MB took 1.046 seconds,   47.800 MB/s   |	SHA-256            605 MB took 1.000 seconds,  604.809 MB/s
SHA-384             85 MB took 1.052 seconds,   80.778 MB/s   |	SHA-384            140 MB took 1.007 seconds,  139.095 MB/s
SHA-512             85 MB took 1.050 seconds,   80.916 MB/s   |	SHA-512            140 MB took 1.005 seconds,  139.309 MB/s
SHA3-224            80 MB took 1.014 seconds,   78.885 MB/s   |	SHA3-224            70 MB took 1.009 seconds,   69.392 MB/s
SHA3-256            75 MB took 1.002 seconds,   74.835 MB/s   |	SHA3-256            70 MB took 1.058 seconds,   66.187 MB/s
SHA3-384            60 MB took 1.028 seconds,   58.372 MB/s   |	SHA3-384            55 MB took 1.037 seconds,   53.043 MB/s
SHA3-512            45 MB took 1.088 seconds,   41.366 MB/s   |	SHA3-512            40 MB took 1.037 seconds,   38.588 MB/s
AES-128-CMAC        30 MB took 1.063 seconds,   28.212 MB/s   |	AES-128-CMAC       170 MB took 1.011 seconds,  168.183 MB/s
AES-256-CMAC        25 MB took 1.144 seconds,   21.854 MB/s   |	AES-256-CMAC       150 MB took 1.023 seconds,  146.610 MB/s
HMAC-MD5           135 MB took 1.018 seconds,  132.559 MB/s   |	HMAC-MD5           135 MB took 1.014 seconds,  133.129 MB/s
HMAC-SHA           115 MB took 1.042 seconds,  110.373 MB/s   |	HMAC-SHA           115 MB took 1.027 seconds,  111.934 MB/s
HMAC-SHA224         50 MB took 1.044 seconds,   47.891 MB/s   |	HMAC-SHA224        610 MB took 1.004 seconds,  607.596 MB/s
HMAC-SHA256         50 MB took 1.044 seconds,   47.884 MB/s   |	HMAC-SHA256        610 MB took 1.005 seconds,  607.016 MB/s
HMAC-SHA384         85 MB took 1.050 seconds,   80.924 MB/s   |	HMAC-SHA384        140 MB took 1.005 seconds,  139.361 MB/s
HMAC-SHA512         85 MB took 1.050 seconds,   80.923 MB/s   |	HMAC-SHA512        140 MB took 1.006 seconds,  139.176 MB/s
PBKDF2               6 KB took 1.004 seconds,    5.573 KB/s   |	PBKDF2              29 KB took 1.001 seconds,   29.229 KB/s
RSA     1024 key gen         4 ops took 1.921 sec, avg 480.21 |	RSA     1024 key gen         3 ops took 1.041 sec, avg 346.98
RSA     2048 key gen         1 ops took 7.526 sec, avg 7525.5 |	RSA     2048 key gen         1 ops took 3.531 sec, avg 3530.8
RSA     2048 public        700 ops took 1.076 sec, avg 1.537  |	RSA     2048 public        700 ops took 1.062 sec, avg 1.517
RSA     2048 private       100 ops took 9.264 sec, avg 92.642 |	RSA     2048 private       100 ops took 9.239 sec, avg 92.392
DH      2048 key gen        58 ops took 1.016 sec, avg 17.523 |	DH      2048 key gen        58 ops took 1.014 sec, avg 17.479
DH      2048 agree         100 ops took 3.813 sec, avg 38.129 |	DH      2048 agree         100 ops took 3.803 sec, avg 38.032
ECC   [      SECP256R1]   256 key gen       116 ops took 1.00 |	ECC   [      SECP256R1]   256 key gen       115 ops took 1.00
ECDHE [      SECP256R1]   256 agree         200 ops took 1.72 |	ECDHE [      SECP256R1]   256 agree         200 ops took 1.73
ECDSA [      SECP256R1]   256 sign          200 ops took 1.77 |	ECDSA [      SECP256R1]   256 sign          200 ops took 1.78
ECDSA [      SECP256R1]   256 verify        200 ops took 1.17 |	ECDSA [      SECP256R1]   256 verify        200 ops took 1.21
CURVE  25519 key gen      2032 ops took 1.000 sec, avg 0.492  |	CURVE  25519 key gen      2940 ops took 1.000 sec, avg 0.340
CURVE  25519 agree        2100 ops took 1.033 sec, avg 0.492  |	CURVE  25519 agree        3000 ops took 1.022 sec, avg 0.341
Benchmark complete     	       	       	       	       	       	Benchmark complete
```

cotequeiroz committed:
wolfssl: add benchmark utility

This packages the wolfssl benchmark utility.

Signed-off-by: Eneas U de Queiroz <cotequeiroz@gmail.com>

robimarko commented:
Ok, so the Octeon TX CN81XX and CN83XX datasheets are a bit cryptic but they do support Crypto as well as CRC extensions.

MeIsReallyBa commented:
Can openssl also support arm ce?

robimarko commented:
Its already enabled in OpenSSL

robimarko commented:
Both rk3399 and rk3328 have the extensions according to the datasheets:
https://www.olimex.com/Products/SOM/RK3328/_resources/RK3328-datasheet-V1.2.pdf
https://www.rockchip.fr/RK3399%20datasheet%20V1.8.pdf

robimarko commented:
It definitively works on Ebin Ultra (Mvebu A53):
```
-------------------------------------------------------------	-------------------------------------------------------------
 wolfSSL version 5.2.0						 wolfSSL version 5.2.0
-------------------------------------------------------------	-------------------------------------------------------------
wolfCrypt Benchmark (block bytes 1048576, min 1.0 sec each)	wolfCrypt Benchmark (block bytes 1048576, min 1.0 sec each)
RNG                 20 MB took 1.316 seconds,   15.193 MB/s   |	RNG                 65 MB took 1.035 seconds,   62.789 MB/s
AES-128-CBC-enc     25 MB took 1.123 seconds,   22.261 MB/s   |	AES-128-CBC-enc    515 MB took 1.006 seconds,  511.678 MB/s
AES-128-CBC-dec     25 MB took 1.216 seconds,   20.556 MB/s   |	AES-128-CBC-dec    480 MB took 1.006 seconds,  477.037 MB/s
AES-192-CBC-enc     20 MB took 1.031 seconds,   19.400 MB/s   |	AES-192-CBC-enc    455 MB took 1.006 seconds,  452.356 MB/s
AES-192-CBC-dec     20 MB took 1.123 seconds,   17.803 MB/s   |	AES-192-CBC-dec    445 MB took 1.008 seconds,  441.670 MB/s
AES-256-CBC-enc     20 MB took 1.163 seconds,   17.195 MB/s   |	AES-256-CBC-enc    410 MB took 1.002 seconds,  409.325 MB/s
AES-256-CBC-dec     20 MB took 1.274 seconds,   15.696 MB/s   |	AES-256-CBC-dec    400 MB took 1.011 seconds,  395.605 MB/s
AES-128-GCM-enc     15 MB took 1.135 seconds,   13.215 MB/s   |	AES-128-GCM-enc    325 MB took 1.009 seconds,  321.980 MB/s
AES-128-GCM-dec     15 MB took 1.135 seconds,   13.214 MB/s   |	AES-128-GCM-dec    155 MB took 1.020 seconds,  151.892 MB/s
AES-192-GCM-enc     15 MB took 1.234 seconds,   12.156 MB/s   |	AES-192-GCM-enc    305 MB took 1.003 seconds,  304.176 MB/s
AES-192-GCM-dec     15 MB took 1.234 seconds,   12.152 MB/s   |	AES-192-GCM-dec    150 MB took 1.030 seconds,  145.662 MB/s
AES-256-GCM-enc     15 MB took 1.334 seconds,   11.248 MB/s   |	AES-256-GCM-enc    285 MB took 1.016 seconds,  280.447 MB/s
AES-256-GCM-dec     15 MB took 1.334 seconds,   11.243 MB/s   |	AES-256-GCM-dec    145 MB took 1.034 seconds,  140.202 MB/s
GMAC Table 4-bit    34 MB took 1.020 seconds,   33.336 MB/s   |	GMAC Table 4-bit   242 MB took 1.000 seconds,  241.903 MB/s
AES-128-ECB-enc     22 MB took 1.000 seconds,   22.298 MB/s   |	AES-128-ECB-enc    156 MB took 1.000 seconds,  155.994 MB/s
AES-128-ECB-dec     21 MB took 1.000 seconds,   20.732 MB/s   |	AES-128-ECB-dec    160 MB took 1.000 seconds,  159.936 MB/s
AES-192-ECB-enc     19 MB took 1.000 seconds,   19.406 MB/s   |	AES-192-ECB-enc    141 MB took 1.000 seconds,  140.981 MB/s
AES-192-ECB-dec     18 MB took 1.000 seconds,   17.938 MB/s   |	AES-192-ECB-dec    145 MB took 1.000 seconds,  144.738 MB/s
AES-256-ECB-enc     17 MB took 1.000 seconds,   17.208 MB/s   |	AES-256-ECB-enc    130 MB took 1.000 seconds,  130.118 MB/s
AES-256-ECB-dec     16 MB took 1.000 seconds,   15.783 MB/s   |	AES-256-ECB-dec    133 MB took 1.000 seconds,  133.349 MB/s
AES-128-CTR         25 MB took 1.150 seconds,   21.731 MB/s   |	AES-128-CTR        650 MB took 1.001 seconds,  649.525 MB/s
AES-192-CTR         20 MB took 1.051 seconds,   19.021 MB/s   |	AES-192-CTR        605 MB took 1.003 seconds,  603.041 MB/s
AES-256-CTR         20 MB took 1.186 seconds,   16.870 MB/s   |	AES-256-CTR        555 MB took 1.000 seconds,  554.806 MB/s
AES-CCM-Enc         15 MB took 1.358 seconds,   11.043 MB/s   |	AES-CCM-Enc         75 MB took 1.004 seconds,   74.713 MB/s
AES-CCM-Dec         15 MB took 1.358 seconds,   11.046 MB/s   |	AES-CCM-Dec         75 MB took 1.004 seconds,   74.665 MB/s
ARC4                50 MB took 1.015 seconds,   49.277 MB/s   |	ARC4                50 MB took 1.021 seconds,   48.978 MB/s
CHACHA              60 MB took 1.049 seconds,   57.173 MB/s   |	CHACHA             220 MB took 1.010 seconds,  217.785 MB/s
CHA-POLY            45 MB took 1.069 seconds,   42.104 MB/s   |	CHA-POLY           150 MB took 1.013 seconds,  148.062 MB/s
3DES                 5 MB took 1.247 seconds,    4.011 MB/s   |	3DES                 5 MB took 1.315 seconds,    3.801 MB/s
MD5                100 MB took 1.010 seconds,   98.979 MB/s   |	MD5                100 MB took 1.013 seconds,   98.729 MB/s
POLY1305           180 MB took 1.021 seconds,  176.236 MB/s   |	POLY1305           455 MB took 1.000 seconds,  454.791 MB/s
SHA                 90 MB took 1.038 seconds,   86.721 MB/s   |	SHA                 85 MB took 1.019 seconds,   83.446 MB/s
SHA-224             40 MB took 1.093 seconds,   36.596 MB/s   |	SHA-224            455 MB took 1.003 seconds,  453.679 MB/s
SHA-256             40 MB took 1.092 seconds,   36.630 MB/s   |	SHA-256            455 MB took 1.001 seconds,  454.514 MB/s
SHA-384             65 MB took 1.071 seconds,   60.704 MB/s   |	SHA-384            105 MB took 1.021 seconds,  102.842 MB/s
SHA-512             65 MB took 1.073 seconds,   60.554 MB/s   |	SHA-512            105 MB took 1.019 seconds,  103.024 MB/s
SHA3-224            60 MB took 1.009 seconds,   59.459 MB/s   |	SHA3-224            55 MB took 1.071 seconds,   51.331 MB/s
SHA3-256            60 MB took 1.062 seconds,   56.513 MB/s   |	SHA3-256            50 MB took 1.020 seconds,   49.030 MB/s
SHA3-384            45 MB took 1.025 seconds,   43.896 MB/s   |	SHA3-384            40 MB took 1.018 seconds,   39.276 MB/s
SHA3-512            35 MB took 1.128 seconds,   31.015 MB/s   |	SHA3-512            30 MB took 1.050 seconds,   28.565 MB/s
AES-128-CMAC        25 MB took 1.168 seconds,   21.406 MB/s   |	AES-128-CMAC       125 MB took 1.015 seconds,  123.110 MB/s
AES-256-CMAC        20 MB took 1.200 seconds,   16.669 MB/s   |	AES-256-CMAC       110 MB took 1.036 seconds,  106.185 MB/s
HMAC-MD5           100 MB took 1.012 seconds,   98.812 MB/s   |	HMAC-MD5           100 MB took 1.012 seconds,   98.808 MB/s
HMAC-SHA            90 MB took 1.040 seconds,   86.500 MB/s   |	HMAC-SHA            85 MB took 1.017 seconds,   83.561 MB/s
HMAC-SHA224         40 MB took 1.094 seconds,   36.547 MB/s   |	HMAC-SHA224        455 MB took 1.006 seconds,  452.312 MB/s
HMAC-SHA256         40 MB took 1.095 seconds,   36.545 MB/s   |	HMAC-SHA256        455 MB took 1.006 seconds,  452.340 MB/s
HMAC-SHA384         65 MB took 1.073 seconds,   60.573 MB/s   |	HMAC-SHA384        105 MB took 1.019 seconds,  103.055 MB/s
HMAC-SHA512         65 MB took 1.071 seconds,   60.663 MB/s   |	HMAC-SHA512        105 MB took 1.018 seconds,  103.111 MB/s
PBKDF2               4 KB took 1.001 seconds,    4.246 KB/s   |	PBKDF2              22 KB took 1.000 seconds,   21.741 KB/s
RSA     1024 key gen         3 ops took 1.699 sec, avg 566.43 |	RSA     1024 key gen         3 ops took 1.517 sec, avg 505.61
RSA     2048 key gen         1 ops took 6.844 sec, avg 6843.6 |	RSA     2048 key gen         1 ops took 5.542 sec, avg 5542.3
RSA     2048 public        500 ops took 1.039 sec, avg 2.079  |	RSA     2048 public        500 ops took 1.035 sec, avg 2.071 
RSA     2048 private       100 ops took 12.516 sec, avg 125.1 |	RSA     2048 private       100 ops took 12.548 sec, avg 125.4
DH      2048 key gen        43 ops took 1.023 sec, avg 23.796 |	DH      2048 key gen        42 ops took 1.001 sec, avg 23.833
DH      2048 agree         100 ops took 5.154 sec, avg 51.536 |	DH      2048 agree         100 ops took 5.162 sec, avg 51.623
ECC   [      SECP256R1]   256 key gen        87 ops took 1.00 |	ECC   [      SECP256R1]   256 key gen        86 ops took 1.00
ECDHE [      SECP256R1]   256 agree         100 ops took 1.15 |	ECDHE [      SECP256R1]   256 agree         100 ops took 1.16
ECDSA [      SECP256R1]   256 sign          100 ops took 1.19 |	ECDSA [      SECP256R1]   256 sign          100 ops took 1.19
ECDSA [      SECP256R1]   256 verify        200 ops took 1.62 |	ECDSA [      SECP256R1]   256 verify        200 ops took 1.60
CURVE  25519 key gen      1510 ops took 1.000 sec, avg 0.662  |	CURVE  25519 key gen      2180 ops took 1.000 sec, avg 0.459 
CURVE  25519 agree        1600 ops took 1.059 sec, avg 0.662  |	CURVE  25519 agree        2200 ops took 1.013 sec, avg 0.460 
Benchmark complete						Benchmark complete
```

mj22226 commented:
I tested nanopi r2s rk3328 and rock pi 4b rk3399 , the difference is noticeable enabling the wolfssl crypto instructions.

rk3328:
```
-------------------------------------------------------------	-------------------------------------------------------------
 wolfSSL version 5.2.0						 wolfSSL version 5.2.0
-------------------------------------------------------------	-------------------------------------------------------------
wolfCrypt Benchmark (block bytes 1048576, min 1.0 sec each)	wolfCrypt Benchmark (block bytes 1048576, min 1.0 sec each)
RNG                 20 MB took 1.037 seconds,   19.292 MB/s   |	RNG                 85 MB took 1.023 seconds,   83.065 MB/s
AES-128-CBC-enc     30 MB took 1.051 seconds,   28.544 MB/s   |	AES-128-CBC-enc    490 MB took 1.002 seconds,  489.049 MB/s
AES-128-CBC-dec     30 MB took 1.145 seconds,   26.209 MB/s   |	AES-128-CBC-dec    450 MB took 1.003 seconds,  448.691 MB/s
AES-192-CBC-enc     25 MB took 1.007 seconds,   24.827 MB/s   |	AES-192-CBC-enc    455 MB took 1.001 seconds,  454.527 MB/s
AES-192-CBC-dec     25 MB took 1.105 seconds,   22.623 MB/s   |	AES-192-CBC-dec    420 MB took 1.004 seconds,  418.418 MB/s
AES-256-CBC-enc     25 MB took 1.137 seconds,   21.997 MB/s   |	AES-256-CBC-enc    440 MB took 1.003 seconds,  438.818 MB/s
AES-256-CBC-dec     20 MB took 1.000 seconds,   19.998 MB/s   |	AES-256-CBC-dec    395 MB took 1.007 seconds,  392.202 MB/s
AES-128-GCM-enc     20 MB took 1.181 seconds,   16.934 MB/s   |	AES-128-GCM-enc    390 MB took 1.011 seconds,  385.873 MB/s
AES-128-GCM-dec     20 MB took 1.180 seconds,   16.945 MB/s   |	AES-128-GCM-dec    185 MB took 1.013 seconds,  182.715 MB/s
AES-192-GCM-enc     20 MB took 1.285 seconds,   15.568 MB/s   |	AES-192-GCM-enc    360 MB took 1.005 seconds,  358.265 MB/s
AES-192-GCM-dec     20 MB took 1.286 seconds,   15.548 MB/s   |	AES-192-GCM-dec    185 MB took 1.009 seconds,  183.331 MB/s
AES-256-GCM-enc     15 MB took 1.044 seconds,   14.367 MB/s   |	AES-256-GCM-enc    340 MB took 1.006 seconds,  337.997 MB/s
AES-256-GCM-dec     15 MB took 1.042 seconds,   14.397 MB/s   |	AES-256-GCM-dec    180 MB took 1.020 seconds,  176.463 MB/s
GMAC Table 4-bit    43 MB took 1.006 seconds,   42.732 MB/s   |	GMAC Table 4-bit   310 MB took 1.001 seconds,  309.764 MB/s
AES-128-ECB-enc     29 MB took 1.000 seconds,   28.824 MB/s   |	AES-128-ECB-enc    202 MB took 1.000 seconds,  202.065 MB/s
AES-128-ECB-dec     27 MB took 1.000 seconds,   26.735 MB/s   |	AES-128-ECB-dec    207 MB took 1.000 seconds,  207.200 MB/s
AES-192-ECB-enc     25 MB took 1.000 seconds,   25.081 MB/s   |	AES-192-ECB-enc    183 MB took 1.000 seconds,  182.630 MB/s
AES-192-ECB-dec     23 MB took 1.000 seconds,   23.107 MB/s   |	AES-192-ECB-dec    188 MB took 1.000 seconds,  187.523 MB/s
AES-256-ECB-enc     22 MB took 1.000 seconds,   22.246 MB/s   |	AES-256-ECB-enc    169 MB took 1.000 seconds,  168.596 MB/s
AES-256-ECB-dec     20 MB took 1.000 seconds,   20.364 MB/s   |	AES-256-ECB-dec    173 MB took 1.000 seconds,  172.751 MB/s
AES-128-CTR         30 MB took 1.073 seconds,   27.951 MB/s   |	AES-128-CTR        555 MB took 1.003 seconds,  553.177 MB/s
AES-192-CTR         25 MB took 1.025 seconds,   24.380 MB/s   |	AES-192-CTR        535 MB took 1.000 seconds,  534.987 MB/s
AES-256-CTR         25 MB took 1.156 seconds,   21.620 MB/s   |	AES-256-CTR        530 MB took 1.001 seconds,  529.458 MB/s
AES-CCM-Enc         15 MB took 1.065 seconds,   14.088 MB/s   |	AES-CCM-Enc        100 MB took 1.049 seconds,   95.285 MB/s
AES-CCM-Dec         15 MB took 1.065 seconds,   14.089 MB/s   |	AES-CCM-Dec        100 MB took 1.032 seconds,   96.887 MB/s
ARC4                65 MB took 1.034 seconds,   62.851 MB/s   |	ARC4                65 MB took 1.037 seconds,   62.662 MB/s
CHACHA              75 MB took 1.042 seconds,   71.947 MB/s   |	CHACHA             270 MB took 1.001 seconds,  269.784 MB/s
CHA-POLY            55 MB took 1.033 seconds,   53.233 MB/s   |	CHA-POLY           190 MB took 1.025 seconds,  185.311 MB/s
3DES                10 MB took 1.969 seconds,    5.078 MB/s   |	3DES                 5 MB took 1.031 seconds,    4.849 MB/s
MD5                130 MB took 1.023 seconds,  127.063 MB/s   |	MD5                130 MB took 1.028 seconds,  126.460 MB/s
POLY1305           230 MB took 1.019 seconds,  225.736 MB/s   |	POLY1305           580 MB took 1.003 seconds,  578.269 MB/s
SHA                115 MB took 1.037 seconds,  110.866 MB/s   |	SHA                110 MB took 1.032 seconds,  106.633 MB/s
SHA-224             50 MB took 1.095 seconds,   45.676 MB/s   |	SHA-224            575 MB took 1.007 seconds,  570.829 MB/s
SHA-256             50 MB took 1.095 seconds,   45.656 MB/s   |	SHA-256            575 MB took 1.003 seconds,  573.153 MB/s
SHA-384             80 MB took 1.026 seconds,   77.951 MB/s   |	SHA-384            135 MB took 1.020 seconds,  132.418 MB/s
SHA-512             80 MB took 1.028 seconds,   77.799 MB/s   |	SHA-512            135 MB took 1.018 seconds,  132.559 MB/s
SHA3-224            80 MB took 1.043 seconds,   76.699 MB/s   |	SHA3-224            70 MB took 1.058 seconds,   66.135 MB/s
SHA3-256            75 MB took 1.031 seconds,   72.770 MB/s   |	SHA3-256            65 MB took 1.029 seconds,   63.174 MB/s
SHA3-384            60 MB took 1.060 seconds,   56.625 MB/s   |	SHA3-384            55 MB took 1.090 seconds,   50.438 MB/s
SHA3-512            40 MB took 1.002 seconds,   39.927 MB/s   |	SHA3-512            40 MB took 1.091 seconds,   36.657 MB/s
AES-128-CMAC        30 MB took 1.107 seconds,   27.098 MB/s   |	AES-128-CMAC       160 MB took 1.019 seconds,  157.010 MB/s
AES-256-CMAC        25 MB took 1.180 seconds,   21.191 MB/s   |	AES-256-CMAC       135 MB took 1.000 seconds,  134.953 MB/s
HMAC-MD5           130 MB took 1.019 seconds,  127.619 MB/s   |	HMAC-MD5           130 MB took 1.024 seconds,  126.921 MB/s
HMAC-SHA           115 MB took 1.044 seconds,  110.154 MB/s   |	HMAC-SHA           110 MB took 1.025 seconds,  107.289 MB/s
HMAC-SHA224         50 MB took 1.097 seconds,   45.560 MB/s   |	HMAC-SHA224        575 MB took 1.000 seconds,  574.753 MB/s
HMAC-SHA256         50 MB took 1.098 seconds,   45.540 MB/s   |	HMAC-SHA256        575 MB took 1.001 seconds,  574.249 MB/s
HMAC-SHA384         80 MB took 1.027 seconds,   77.891 MB/s   |	HMAC-SHA384        135 MB took 1.024 seconds,  131.816 MB/s
HMAC-SHA512         80 MB took 1.028 seconds,   77.814 MB/s   |	HMAC-SHA512        135 MB took 1.020 seconds,  132.375 MB/s
PBKDF2               5 KB took 1.001 seconds,    5.336 KB/s   |	PBKDF2              28 KB took 1.000 seconds,   28.243 KB/s
RSA     1024 key gen         3 ops took 1.234 sec, avg 411.32 |	RSA     1024 key gen         2 ops took 1.439 sec, avg 719.50
RSA     2048 key gen         1 ops took 5.928 sec, avg 5927.8 |	RSA     2048 key gen         1 ops took 1.876 sec, avg 1875.8
RSA     2048 public        700 ops took 1.127 sec, avg 1.610  |	RSA     2048 public        700 ops took 1.120 sec, avg 1.600 
RSA     2048 private       100 ops took 9.703 sec, avg 97.030 |	RSA     2048 private       100 ops took 9.710 sec, avg 97.102
DH      2048 key gen        55 ops took 1.013 sec, avg 18.416 |	DH      2048 key gen        55 ops took 1.014 sec, avg 18.438
DH      2048 agree         100 ops took 3.989 sec, avg 39.887 |	DH      2048 agree         100 ops took 4.073 sec, avg 40.726
ECC   [      SECP256R1]   256 key gen       112 ops took 1.00 |	ECC   [      SECP256R1]   256 key gen       111 ops took 1.00
ECDHE [      SECP256R1]   256 agree         200 ops took 1.79 |	ECDHE [      SECP256R1]   256 agree         200 ops took 1.80
ECDSA [      SECP256R1]   256 sign          200 ops took 1.84 |	ECDSA [      SECP256R1]   256 sign          200 ops took 1.85
ECDSA [      SECP256R1]   256 verify        200 ops took 1.25 |	ECDSA [      SECP256R1]   256 verify        200 ops took 1.29
CURVE  25519 key gen      1959 ops took 1.000 sec, avg 0.511  |	CURVE  25519 key gen      2818 ops took 1.000 sec, avg 0.355 
CURVE  25519 agree        2000 ops took 1.021 sec, avg 0.510  |	CURVE  25519 agree        2900 ops took 1.030 sec, avg 0.355 
Benchmark complete						Benchmark complete
```
rk3399:
```
-------------------------------------------------------------   -------------------------------------------------------------
 wolfSSL version 5.2.0                                           wolfSSL version 5.2.0
-------------------------------------------------------------   -------------------------------------------------------------
wolfCrypt Benchmark (block bytes 1048576, min 1.0 sec each)     wolfCrypt Benchmark (block bytes 1048576, min 1.0 sec each)
RNG                 45 MB took 1.055 seconds,   42.657 MB/s   | RNG                205 MB took 1.021 seconds,  200.720 MB/s
AES-128-CBC-enc     85 MB took 1.022 seconds,   83.140 MB/s   | AES-128-CBC-enc   1000 MB took 1.003 seconds,  996.759 MB/s
AES-128-CBC-dec     85 MB took 1.039 seconds,   81.828 MB/s   | AES-128-CBC-dec   1520 MB took 1.002 seconds, 1517.215 MB/s
AES-192-CBC-enc     75 MB took 1.040 seconds,   72.136 MB/s   | AES-192-CBC-enc    870 MB took 1.002 seconds,  868.419 MB/s
AES-192-CBC-dec     75 MB took 1.057 seconds,   70.944 MB/s   | AES-192-CBC-dec   1355 MB took 1.001 seconds, 1352.978 MB/s
AES-256-CBC-enc     65 MB took 1.025 seconds,   63.397 MB/s   | AES-256-CBC-enc    775 MB took 1.005 seconds,  771.512 MB/s
AES-256-CBC-dec     65 MB took 1.037 seconds,   62.668 MB/s   | AES-256-CBC-dec   1105 MB took 1.000 seconds, 1104.481 MB/s
AES-128-GCM-enc     50 MB took 1.083 seconds,   46.178 MB/s   | AES-128-GCM-enc    595 MB took 1.003 seconds,  592.979 MB/s
AES-128-GCM-dec     50 MB took 1.083 seconds,   46.166 MB/s   | AES-128-GCM-dec    315 MB took 1.006 seconds,  313.136 MB/s
AES-192-GCM-enc     45 MB took 1.057 seconds,   42.588 MB/s   | AES-192-GCM-enc    590 MB took 1.005 seconds,  587.107 MB/s
AES-192-GCM-dec     45 MB took 1.057 seconds,   42.566 MB/s   | AES-192-GCM-dec    300 MB took 1.002 seconds,  299.350 MB/s
AES-256-GCM-enc     40 MB took 1.017 seconds,   39.314 MB/s   | AES-256-GCM-enc    545 MB took 1.006 seconds,  541.682 MB/s
AES-256-GCM-dec     40 MB took 1.017 seconds,   39.338 MB/s   | AES-256-GCM-dec    290 MB took 1.010 seconds,  286.988 MB/s
GMAC Table 4-bit   107 MB took 1.005 seconds,  106.511 MB/s   | GMAC Table 4-bit   497 MB took 1.000 seconds,  496.993 MB/s
AES-128-ECB-enc     80 MB took 1.000 seconds,   79.806 MB/s   | AES-128-ECB-enc    497 MB took 1.000 seconds,  497.402 MB/s
AES-128-ECB-dec     79 MB took 1.000 seconds,   79.421 MB/s   | AES-128-ECB-dec    522 MB took 1.000 seconds,  521.899 MB/s
AES-192-ECB-enc     70 MB took 1.000 seconds,   69.712 MB/s   | AES-192-ECB-enc    452 MB took 1.000 seconds,  451.590 MB/s
AES-192-ECB-dec     67 MB took 1.000 seconds,   67.478 MB/s   | AES-192-ECB-dec    474 MB took 1.000 seconds,  474.384 MB/s
AES-256-ECB-enc     62 MB took 1.000 seconds,   61.505 MB/s   | AES-256-ECB-enc    407 MB took 1.000 seconds,  407.336 MB/s
AES-256-ECB-dec     60 MB took 1.000 seconds,   59.630 MB/s   | AES-256-ECB-dec    433 MB took 1.000 seconds,  433.024 MB/s
AES-128-CTR         85 MB took 1.041 seconds,   81.686 MB/s   | AES-128-CTR       1470 MB took 1.001 seconds, 1468.157 MB/s
AES-192-CTR         75 MB took 1.055 seconds,   71.077 MB/s   | AES-192-CTR       1350 MB took 1.001 seconds, 1348.042 MB/s
AES-256-CTR         65 MB took 1.041 seconds,   62.436 MB/s   | AES-256-CTR       1235 MB took 1.000 seconds, 1234.821 MB/s
AES-CCM-Enc         45 MB took 1.095 seconds,   41.081 MB/s   | AES-CCM-Enc        235 MB took 1.000 seconds,  234.950 MB/s
AES-CCM-Dec         45 MB took 1.096 seconds,   41.042 MB/s   | AES-CCM-Dec        230 MB took 1.014 seconds,  226.790 MB/s
ARC4               170 MB took 1.007 seconds,  168.763 MB/s   | ARC4               170 MB took 1.006 seconds,  169.001 MB/s
CHACHA             145 MB took 1.008 seconds,  143.837 MB/s   | CHACHA             425 MB took 1.008 seconds,  421.673 MB/s
CHA-POLY           105 MB took 1.013 seconds,  103.606 MB/s   | CHA-POLY           280 MB took 1.003 seconds,  279.203 MB/s
3DES                15 MB took 1.076 seconds,   13.945 MB/s   | 3DES                15 MB took 1.128 seconds,   13.297 MB/s
MD5                270 MB took 1.002 seconds,  269.415 MB/s   | MD5                270 MB took 1.010 seconds,  267.258 MB/s
POLY1305           370 MB took 1.007 seconds,  367.540 MB/s   | POLY1305           825 MB took 1.000 seconds,  824.983 MB/s
SHA                200 MB took 1.006 seconds,  198.816 MB/s   | SHA                200 MB took 1.010 seconds,  197.969 MB/s
SHA-224            100 MB took 1.019 seconds,   98.160 MB/s   | SHA-224            930 MB took 1.003 seconds,  927.238 MB/s
SHA-256            100 MB took 1.019 seconds,   98.151 MB/s   | SHA-256            930 MB took 1.001 seconds,  929.247 MB/s
SHA-384            180 MB took 1.006 seconds,  178.875 MB/s   | SHA-384            190 MB took 1.002 seconds,  189.713 MB/s
SHA-512            180 MB took 1.006 seconds,  178.997 MB/s   | SHA-512            190 MB took 1.001 seconds,  189.748 MB/s
SHA3-224           155 MB took 1.010 seconds,  153.530 MB/s   | SHA3-224           145 MB took 1.022 seconds,  141.854 MB/s
SHA3-256           145 MB took 1.009 seconds,  143.742 MB/s   | SHA3-256           135 MB took 1.001 seconds,  134.817 MB/s
SHA3-384           115 MB took 1.006 seconds,  114.341 MB/s   | SHA3-384           110 MB took 1.027 seconds,  107.072 MB/s
SHA3-512            85 MB took 1.057 seconds,   80.428 MB/s   | SHA3-512            80 MB took 1.047 seconds,   76.411 MB/s
AES-128-CMAC        80 MB took 1.005 seconds,   79.623 MB/s   | AES-128-CMAC       515 MB took 1.002 seconds,  514.060 MB/s
AES-256-CMAC        65 MB took 1.060 seconds,   61.313 MB/s   | AES-256-CMAC       435 MB took 1.006 seconds,  432.611 MB/s
HMAC-MD5           270 MB took 1.013 seconds,  266.566 MB/s   | HMAC-MD5           270 MB took 1.011 seconds,  266.961 MB/s
HMAC-SHA           200 MB took 1.002 seconds,  199.538 MB/s   | HMAC-SHA           195 MB took 1.001 seconds,  194.898 MB/s
HMAC-SHA224        100 MB took 1.019 seconds,   98.158 MB/s   | HMAC-SHA224        930 MB took 1.001 seconds,  929.404 MB/s
HMAC-SHA256        100 MB took 1.019 seconds,   98.153 MB/s   | HMAC-SHA256        935 MB took 1.005 seconds,  930.703 MB/s
HMAC-SHA384        180 MB took 1.006 seconds,  179.003 MB/s   | HMAC-SHA384        190 MB took 1.001 seconds,  189.782 MB/s
HMAC-SHA512        180 MB took 1.006 seconds,  179.012 MB/s   | HMAC-SHA512        190 MB took 1.001 seconds,  189.786 MB/s
PBKDF2              12 KB took 1.002 seconds,   11.607 KB/s   | PBKDF2              63 KB took 1.000 seconds,   63.059 KB/s
RSA     1024 key gen         5 ops took 1.047 sec, avg 209.43 | RSA     1024 key gen         6 ops took 1.049 sec, avg 174.86
RSA     2048 key gen         1 ops took 1.123 sec, avg 1122.7 | RSA     2048 key gen         2 ops took 1.636 sec, avg 817.76
RSA     2048 public       1600 ops took 1.048 sec, avg 0.655  | RSA     2048 public       1600 ops took 1.041 sec, avg 0.651
RSA     2048 private       100 ops took 3.665 sec, avg 36.646 | RSA     2048 private       100 ops took 3.659 sec, avg 36.591
DH      2048 key gen       131 ops took 1.002 sec, avg 7.649  | DH      2048 key gen       131 ops took 1.005 sec, avg 7.672
DH      2048 agree         100 ops took 1.500 sec, avg 14.999 | DH      2048 agree         100 ops took 1.501 sec, avg 15.014
ECC   [      SECP256R1]   256 key gen       235 ops took 1.00 | ECC   [      SECP256R1]   256 key gen       236 ops took 1.00
ECDHE [      SECP256R1]   256 agree         300 ops took 1.28 | ECDHE [      SECP256R1]   256 agree         300 ops took 1.27
ECDSA [      SECP256R1]   256 sign          300 ops took 1.31 | ECDSA [      SECP256R1]   256 sign          300 ops took 1.30
ECDSA [      SECP256R1]   256 verify        400 ops took 1.17 | ECDSA [      SECP256R1]   256 verify        400 ops took 1.17
CURVE  25519 key gen      3273 ops took 1.000 sec, avg 0.306  | CURVE  25519 key gen      4917 ops took 1.000 sec, avg 0.203
CURVE  25519 agree        3300 ops took 1.000 sec, avg 0.303  | CURVE  25519 agree        5000 ops took 1.013 sec, avg 0.203
Benchmark complete                                              Benchmark complete
```

robimarko commented:
Mvebu A72 works as well:
```
-------------------------------------------------------------	-------------------------------------------------------------
 wolfSSL version 5.2.0						 wolfSSL version 5.2.0
-------------------------------------------------------------	-------------------------------------------------------------
wolfCrypt Benchmark (block bytes 1048576, min 1.0 sec each)	wolfCrypt Benchmark (block bytes 1048576, min 1.0 sec each)
RNG                 35 MB took 1.053 seconds,   33.224 MB/s   |	RNG                160 MB took 1.024 seconds,  156.278 MB/s
AES-128-CBC-enc     65 MB took 1.001 seconds,   64.947 MB/s   |	AES-128-CBC-enc    790 MB took 1.002 seconds,  788.735 MB/s
AES-128-CBC-dec     65 MB took 1.005 seconds,   64.658 MB/s   |	AES-128-CBC-dec   1215 MB took 1.004 seconds, 1210.574 MB/s
AES-192-CBC-enc     60 MB took 1.064 seconds,   56.385 MB/s   |	AES-192-CBC-enc    685 MB took 1.003 seconds,  682.956 MB/s
AES-192-CBC-dec     60 MB took 1.074 seconds,   55.866 MB/s   |	AES-192-CBC-dec   1075 MB took 1.002 seconds, 1072.801 MB/s
AES-256-CBC-enc     50 MB took 1.011 seconds,   49.474 MB/s   |	AES-256-CBC-enc    615 MB took 1.007 seconds,  610.542 MB/s
AES-256-CBC-dec     50 MB took 1.014 seconds,   49.294 MB/s   |	AES-256-CBC-dec    880 MB took 1.001 seconds,  878.978 MB/s
AES-128-GCM-enc     40 MB took 1.111 seconds,   36.002 MB/s   |	AES-128-GCM-enc    470 MB took 1.006 seconds,  467.101 MB/s
AES-128-GCM-dec     40 MB took 1.110 seconds,   36.025 MB/s   |	AES-128-GCM-dec    245 MB took 1.012 seconds,  242.076 MB/s
AES-192-GCM-enc     35 MB took 1.054 seconds,   33.216 MB/s   |	AES-192-GCM-enc    465 MB took 1.007 seconds,  461.648 MB/s
AES-192-GCM-dec     35 MB took 1.053 seconds,   33.223 MB/s   |	AES-192-GCM-dec    235 MB took 1.013 seconds,  231.877 MB/s
AES-256-GCM-enc     35 MB took 1.140 seconds,   30.691 MB/s   |	AES-256-GCM-enc    425 MB took 1.001 seconds,  424.403 MB/s
AES-256-GCM-dec     35 MB took 1.140 seconds,   30.693 MB/s   |	AES-256-GCM-dec    220 MB took 1.000 seconds,  219.982 MB/s
GMAC Table 4-bit    84 MB took 1.011 seconds,   83.119 MB/s   |	GMAC Table 4-bit   390 MB took 1.002 seconds,  389.276 MB/s
AES-128-ECB-enc     62 MB took 1.000 seconds,   61.556 MB/s   |	AES-128-ECB-enc    388 MB took 1.000 seconds,  387.828 MB/s
AES-128-ECB-dec     62 MB took 1.000 seconds,   62.028 MB/s   |	AES-128-ECB-dec    404 MB took 1.000 seconds,  404.163 MB/s
AES-192-ECB-enc     53 MB took 1.000 seconds,   53.444 MB/s   |	AES-192-ECB-enc    352 MB took 1.000 seconds,  352.305 MB/s
AES-192-ECB-dec     54 MB took 1.000 seconds,   53.916 MB/s   |	AES-192-ECB-dec    368 MB took 1.000 seconds,  367.962 MB/s
AES-256-ECB-enc     48 MB took 1.000 seconds,   47.537 MB/s   |	AES-256-ECB-enc    312 MB took 1.000 seconds,  312.291 MB/s
AES-256-ECB-dec     48 MB took 1.000 seconds,   47.766 MB/s   |	AES-256-ECB-dec    338 MB took 1.000 seconds,  337.806 MB/s
AES-128-CTR         65 MB took 1.019 seconds,   63.778 MB/s   |	AES-128-CTR       1205 MB took 1.001 seconds, 1204.340 MB/s
AES-192-CTR         60 MB took 1.081 seconds,   55.494 MB/s   |	AES-192-CTR       1075 MB took 1.004 seconds, 1071.156 MB/s
AES-256-CTR         50 MB took 1.025 seconds,   48.778 MB/s   |	AES-256-CTR        995 MB took 1.002 seconds,  993.394 MB/s
AES-CCM-Enc         35 MB took 1.092 seconds,   32.061 MB/s   |	AES-CCM-Enc        180 MB took 1.023 seconds,  175.934 MB/s
AES-CCM-Dec         35 MB took 1.094 seconds,   32.002 MB/s   |	AES-CCM-Dec        175 MB took 1.012 seconds,  172.976 MB/s
ARC4               135 MB took 1.027 seconds,  131.429 MB/s   |	ARC4               135 MB took 1.024 seconds,  131.841 MB/s
CHACHA             115 MB took 1.028 seconds,  111.920 MB/s   |	CHACHA             330 MB took 1.003 seconds,  328.922 MB/s
CHA-POLY            85 MB took 1.054 seconds,   80.681 MB/s   |	CHA-POLY           220 MB took 1.010 seconds,  217.896 MB/s
3DES                15 MB took 1.382 seconds,   10.853 MB/s   |	3DES                15 MB took 1.446 seconds,   10.372 MB/s
MD5                210 MB took 1.006 seconds,  208.723 MB/s   |	MD5                210 MB took 1.006 seconds,  208.795 MB/s
POLY1305           295 MB took 1.011 seconds,  291.773 MB/s   |	POLY1305           650 MB took 1.003 seconds,  648.084 MB/s
SHA                160 MB took 1.027 seconds,  155.749 MB/s   |	SHA                155 MB took 1.006 seconds,  154.133 MB/s
SHA-224             80 MB took 1.046 seconds,   76.517 MB/s   |	SHA-224            730 MB took 1.005 seconds,  726.565 MB/s
SHA-256             80 MB took 1.046 seconds,   76.500 MB/s   |	SHA-256            730 MB took 1.005 seconds,  726.234 MB/s
SHA-384            145 MB took 1.027 seconds,  141.124 MB/s   |	SHA-384            150 MB took 1.011 seconds,  148.304 MB/s
SHA-512            145 MB took 1.027 seconds,  141.124 MB/s   |	SHA-512            150 MB took 1.011 seconds,  148.308 MB/s
SHA3-224           120 MB took 1.004 seconds,  119.563 MB/s   |	SHA3-224           115 MB took 1.039 seconds,  110.662 MB/s
SHA3-256           115 MB took 1.014 seconds,  113.450 MB/s   |	SHA3-256           110 MB took 1.045 seconds,  105.252 MB/s
SHA3-384            90 MB took 1.009 seconds,   89.155 MB/s   |	SHA3-384            85 MB took 1.018 seconds,   83.512 MB/s
SHA3-512            65 MB took 1.037 seconds,   62.655 MB/s   |	SHA3-512            60 MB took 1.009 seconds,   59.441 MB/s
AES-128-CMAC        60 MB took 1.009 seconds,   59.468 MB/s   |	AES-128-CMAC       405 MB took 1.009 seconds,  401.323 MB/s
AES-256-CMAC        50 MB took 1.082 seconds,   46.232 MB/s   |	AES-256-CMAC       340 MB took 1.007 seconds,  337.599 MB/s
HMAC-MD5           210 MB took 1.009 seconds,  208.126 MB/s   |	HMAC-MD5           210 MB took 1.015 seconds,  206.989 MB/s
HMAC-SHA           160 MB took 1.027 seconds,  155.731 MB/s   |	HMAC-SHA           155 MB took 1.016 seconds,  152.490 MB/s
HMAC-SHA224         80 MB took 1.046 seconds,   76.506 MB/s   |	HMAC-SHA224        730 MB took 1.005 seconds,  726.204 MB/s
HMAC-SHA256         80 MB took 1.046 seconds,   76.509 MB/s   |	HMAC-SHA256        730 MB took 1.007 seconds,  725.091 MB/s
HMAC-SHA384        145 MB took 1.025 seconds,  141.437 MB/s   |	HMAC-SHA384        150 MB took 1.013 seconds,  148.015 MB/s
HMAC-SHA512        145 MB took 1.025 seconds,  141.434 MB/s   |	HMAC-SHA512        150 MB took 1.013 seconds,  148.055 MB/s
PBKDF2               9 KB took 1.001 seconds,    8.958 KB/s   |	PBKDF2              49 KB took 1.001 seconds,   49.161 KB/s
RSA     1024 key gen         2 ops took 1.076 sec, avg 537.95 |	RSA     1024 key gen         5 ops took 1.051 sec, avg 210.11
RSA     2048 key gen         1 ops took 1.759 sec, avg 1758.8 |	RSA     2048 key gen         2 ops took 4.423 sec, avg 2211.3
RSA     2048 public       1200 ops took 1.011 sec, avg 0.843  |	RSA     2048 public       1200 ops took 1.005 sec, avg 0.837 
RSA     2048 private       100 ops took 4.710 sec, avg 47.097 |	RSA     2048 private       100 ops took 4.700 sec, avg 47.003
DH      2048 key gen       102 ops took 1.004 sec, avg 9.847  |	DH      2048 key gen       102 ops took 1.007 sec, avg 9.872 
DH      2048 agree         100 ops took 1.929 sec, avg 19.288 |	DH      2048 agree         100 ops took 1.929 sec, avg 19.289
ECC   [      SECP256R1]   256 key gen       180 ops took 1.00 |	ECC   [      SECP256R1]   256 key gen       182 ops took 1.00
ECDHE [      SECP256R1]   256 agree         200 ops took 1.11 |	ECDHE [      SECP256R1]   256 agree         200 ops took 1.09
ECDSA [      SECP256R1]   256 sign          200 ops took 1.14 |	ECDSA [      SECP256R1]   256 sign          200 ops took 1.12
ECDSA [      SECP256R1]   256 verify        300 ops took 1.17 |	ECDSA [      SECP256R1]   256 verify        300 ops took 1.14
CURVE  25519 key gen      2547 ops took 1.000 sec, avg 0.393  |	CURVE  25519 key gen      3827 ops took 1.000 sec, avg 0.261 
CURVE  25519 agree        2600 ops took 1.013 sec, avg 0.390  |	CURVE  25519 agree        3900 ops took 1.016 sec, avg 0.260 
Benchmark complete						Benchmark complete
```

cotequeiroz commented:
I added a temporary benchmark package, so someone, hopefully, can test whether enabling armv8-CE breaks `bcm271[01]` or not.  I could not find any CPU detection code, so I simply assume they would not work.  I really hope to be wrong here, as it would make things much easier.

Also, the CPU instructions are treated by WolfSSL the same way as hardware crypto.  Different options don't coexist well.  If you build with more than one, you get redefined functions.

I strongly urge people against using cryptodev or afalg here or in openssl.  I haven't seen a useful case for them.  Using external hardware in kernel space (VPN, disk encryption) is good; it speeds up things.  In userspace (openssl, wolfssl), the context switches will ruin any possible gain you may have.  It's like not having a speed limit in a road, you can go as fast as you can, but then you have to stop at every intersection for 1 minute.

robimarko commented:
Pi BCM targets don't have CE at all, it's not mentioned in the specifications, and Googling around comes to the conclusion that they saved some money by not purchasing the optional CE license.

cotequeiroz commented:
It’s not about the inclusion of CE or not, but whether WolfSSL has a fallback or not. I’m not seeing it, but I have a hard time following their code.

robimarko commented:
Understood, I don't see a fallback in the code, it will most likely fail with a ton of illegal instruction exceptions as the ASM optimized code just checks whether its AARCH64 and is ARMASM enabled, no runtime detection or feature checks

cotequeiroz committed:
wolfssl: don't change ABI because of hw crypto

Enabling different hardware crypto acceleration should not change the
library ABI.  Add them to PKG_CONFIG_DEPENDS after the ABI version hash
has been computed.

Signed-off-by: Eneas U de Queiroz <cotequeiroz@gmail.com>

robimarko commented:
@cotequeiroz So, we are waiting on the Pi target test?
If so, I can drag out a Pi4 over the weekend

cotequeiroz commented:
It would be great.  In the extremely unlikely chance that they work, then this is a no-brainer.

I just don't know what is the best course of action with regards to the remaining untested targets.  Should I just leave them out, or count on the other information we have that point to CE support?  🤔 
I would be more aggressive and enable it based on the evidence that we have.  Should this be proven wrong, I'm sure somebody will scream at me, but we already know how to fix it.
Meanwhile, I've marked #9756 as ready for review.  Consequences of a miss there are minimal.

cotequeiroz commented:
I've finally tested in with the Pi4, and got the expected failure:
```
root@b02:/tmp/wolfssl/wolfssl-benchmark# ./benchmark
------------------------------------------------------------------------------
 wolfSSL version 5.2.0
------------------------------------------------------------------------------
wolfCrypt Benchmark (block bytes 1048576, min 1.0 sec each)
Illegal instruction
```
This is ready to go, based on the evidence that we go so far.  @mans0n, please remove the RFT and WIP labels.

robimarko commented:
Awesome, nice to see this moving

hauke commented on package/libs/wolfssl/Makefile:
```diff
@@ -25,25 +26,42 @@ PKG_MAINTAINER:=Eneas U de Queiroz <cotequeiroz@gmail.com>
+define Package/libwolfssl
+$(call Package/libwolfssl/Default)
+  TITLE:=wolfSSL library
+  PKGFLAGS:=nonshared
```
Is `PKGFLAGS` needed when you already set `PKG_FLAGS:=nonshared`

cotequeiroz replied:
The `PKG_FLAGS` was mistakenly left behind when I inverted the commit ordering.  I've fixed this now.

hauke commented:
If someone backports this to a release branch we should do this shortly before a new minor release because changing a package to `nonshared` will remove the package from the shared packages, but the nonshared are only build in a release.

This pull request looks fine to me.

cotequeiroz commented:
Thanks for the review.

Ansuel commented:
@hauke Any news on this? Looks ready to me aside the untested targets.

cotequeiroz commented:
While at it, let's not forget about #9756, which does the same thing to the kernel, with the added simplicity of not breaking anything that does not support CE.

Ansuel commented:
@hauke @cotequeiroz i merged the kernel part now we need to understand how to handle the `nonshared` problem.

Ansuel commented:
@cotequeiroz think i will merge this on master and create another pr for 22.03?

cotequeiroz commented:
Yes.

cotequeiroz commented:
It seems that the phase2 bots are picking up a different library than the one built by phase1.  Phase1 builds `libwolfssl5.3.0.19fa215e`.
However, the packages in phase2 all have dependencies to `libwolfssl5.3.0.ee39414e`.
The images appear to be correct, but then `opkg` will not work because of the different versions.
```
Collected errors:
 * pkg_hash_check_unresolved: cannot find dependency libwolfssl5.3.0.ee39414e for libustream-wolfssl20201210
 * pkg_hash_fetch_best_installation_candidate: Packages for libustream-wolfssl found, but incompatible with the architectures configured
 * opkg_install_cmd: Cannot install package libustream-wolfssl.
 * pkg_hash_check_unresolved: cannot find dependency libwolfssl5.3.0.ee39414e for px5g-wolfssl
 * pkg_hash_fetch_best_installation_candidate: Packages for px5g-wolfssl found, but incompatible with the architectures configured
 * satisfy_dependencies_for: Cannot satisfy the following dependencies for luci-ssl:
 * 	libwolfssl5.3.0.ee39414e
 ```
  
 ```
$ wget -q 'https://downloads.openwrt.org/snapshots/packages/mips_4kec/base/libwolfssl-benchmark_5.3.0-stable-4_mips_4kec.ipk' -O - | tar xzf - -O ./control.tar.gz | tar xzf - -O ./control
Package: libwolfssl-benchmark
Version: 5.3.0-stable-4
Depends: libc, libwolfssl5.3.0.ee39414e
Source: feeds/base/package/libs/wolfssl
SourceName: libwolfssl-benchmark
License: GPL-2.0-or-later
LicenseFiles: LICENSING COPYING
Section: libs
SourceDateEpoch: 1654592209
CPE-ID: cpe:/a:wolfssl:wolfssl
Maintainer: Eneas U de Queiroz <cotequeiroz@gmail.com>
Architecture: mips_4kec
Installed-Size: 19289
Description:  This is the wolfssl benchmark utility.
```

Ansuel commented:
@cotequeiroz mhhhhhhh is this a bug with opkg ? should we revert?

cotequeiroz commented:
It's definitely not a bug with opkg!

We need to figure out what the bots are doing exactly.  
It may be because the `libwolfssl` package is built by phase1, since it is non-shared, and then when phase2 tries to build the benchmark utility it may build a different version of the library package as a side-effect.  If this be the case, then we can make both packages non-shared until we come up with a better solution.  A revert is always an option.

Or it may be that there's still an old library lurking around somewhere that needs to be cleaned up.

Ansuel commented:
@cotequeiroz could be the benchmark package... I mean a quick workaround would be remove the benchmark package and watch the effects (or flagging it non-shared)

but it's strange that the packet is getting overwritten...

Ansuel commented:
I mean the fix would be to make pahse2 skip non-shared package since they are already built by phase1

Ansuel commented:
I open an issue and set the PKGFLAGS global since wolfssl is a pretty critical package... i think we already have enough info the track this problem

cotequeiroz commented:
Please revert this.  I think the problem is happening because of different configuration options between phase1 and phase2.

Ansuel commented:
@cotequeiroz wouldn't setting the nonshared global indirectly fix that? Or the problem here is that wolfssl is getting recompiled on pahse2?

cotequeiroz commented:
`CONFIG_WOLFSSL_HAS_OPENVPN` is forced to `y` when `CONFIG_ALL=y`.  I think this is the cause.  We can set it to `y` by default, and then the hashes should match.  Let me check it quickly.

cotequeiroz commented:
I can confirm this.  I've got both hashes by running a default config, and then a default config + `CONFIG_ALL=y`.  It should be easy enough:
``` patch
--- a/package/libs/wolfssl/Config.in
+++ b/package/libs/wolfssl/Config.in
@@ -53,7 +53,7 @@ config WOLFSSL_HAS_ECC25519
 
 config WOLFSSL_HAS_OPENVPN
 	bool "Include OpenVPN support"
-	default n
+	default y
 
 config WOLFSSL_ALT_NAMES
 	bool "Include SAN (Subject Alternative Name) support"
```

Ansuel commented:
Mhh if you want i can create make the change and revert the wrong hack.

Ansuel commented:
@cotequeiroz tell me when you are good and i will merge :D

cotequeiroz commented:
Revert the last nonshared commit, and then merge https://github.com/cotequeiroz/openwrt-1/commit/ef4eb2c3288f3de63a7f0dd8a43f99e85b4eeb43

Ansuel commented:
@cotequeiroz @jow- just reported that his cause crashed with x86 for invalid opcode
looks like some cpu have AES-NI but not SSE4 instructions...

Any idea of this? I think reverting would be the quick solution but it would be good if we fix this without reverting the non-shared thing
